discover.py

```diff
@@ -0,0 +1,234 @@
+                stop_dnsmasq()
+                self.last_dnsmasq_command = ''
+            return
+
+        command = get_dns_conf_string(interface)
+        if command == self.last_dnsmasq_command:
+            return
+
+        rewrite_dnsmasq_conf(app.config['DNS_CONF_FILE'], command)
+        reload_dnsmasq()
+        self.last_dnsmasq_command = command
+
+
+def get_dns_conf_string(interface):
+    ip = get_ipv4_addr(interface)
+    common_args = [
+        'domain-needed',
+        'bogus-priv',
+        'server=208.67.222.222',
+        'server=208.67.220.220',
+        'no-resolv',
+        'no-hosts',
+        'bind-interfaces',
+    ]
+    listen_args = [
+        'interface={}'.format(interface),
+        'listen-address={}'.format(ip),
+    ]
+
+    address_records = [
+        (node['hostname'], node['address'])
+        for node in nodes.get(interface, {}).values()
+        if not node['is_local']
+    ]
+    address_records.sort()
+    address_records.insert(0, (app.config['LIQUID_DOMAIN'], ip))
+    address_args = [
+        'address=/{}/{}'.format(*record)
+        for record in address_records
+    ]
+
+    args = common_args + listen_args + address_args
+    return "\n".join(args) + "\n"
+
+def dict_decode(data):
+    def decode(data):
+        if isinstance(data, bytes):
+            return data.decode("latin-1")
+        elif isinstance(data, str):
+            return data
+        return str(data)
+
+    return {
+        decode(key): decode(data[key])
+        for key in data
+    }
+
+def get_ipv4_addr(interface):
+    addrs = netifaces.ifaddresses(interface).get(netifaces.AF_INET)
+    if not addrs:
+        return None
+
+    ips = [
+        a['addr']
+        for a in addrs
+        if 'addr' in a and a.get('netmask') != HOST_ONLY_NETWORK_MASK
+    ]
+
+    if ips:
+        assert len(ips) == 1, "Only one ip per interface is supported."
+        return ips[0]
+    else:
+        return None
+
+def handle_new_service(name, interface, service_info):
+    properties = dict_decode(service_info.properties)
+    hostname = properties.get('liquid_hostname')
+    if not hostname:
+        return
+    is_local = (hostname == app.config['LIQUID_DOMAIN'])
+    if is_local:
+        address = get_ipv4_addr(interface)
+    else:
+        address = ".".join(str(x) for x in service_info.address)
+
+    data = {
+        "hostname": hostname,
+        "is_local": is_local,
+        "address": address,
+        "discovered_at": datetime.now().isoformat()
+    }
+    add_record(name, interface, data)
+
+def add_record(name, interface, data):
+    nodes.setdefault(interface, {})
+    nodes[interface][name] = data
+
+def remove_record(name, interface):
+    del nodes[interface][name]
+
+class LiquidServiceListener(object):
+    def __init__(self, interface):
+        self.interface = interface
+
+    def add_service(self, zeroconf, type_, name):
+        log.debug("+ %s add_service of type %s and name %s", self.interface, type_, name)
```

Comment:
You can use `%r` instead of `%s` so the strings are quoted which makes the messages more readable.